src/shared/discovery.js

```diff
@@ -0,0 +1,204 @@
+'use strict';
+
+/**
+ * Callback invoked when another frame is discovered in this window which runs
+ * the Hypothesis sidebar or annotation layer code.
+ *
+ * @type {Function} DiscoveryCallback
+ * @param {Window} source - The frame that was discovered.
+ * @param {string} origin - The origin to use when posting messages to this frame.
+ * @param {string} token - A random identifier used by this frame.
+ */
+
+/**
+ * Discovery finds frames in the current tab/window that can be annotated (the
+ * "clients") or can fetch annotations from the backend (the "server").
+ *
+ * Currently only one frame can be designated as the server.
+ * (FIXME: This causes problems. See https://github.com/hypothesis/client/issues/249,
+ * https://github.com/hypothesis/client/issues/187).
+ */
+class Discovery {
+  /**
+   * @param {Window} target
+   * @param {Object} options
+   */
+  constructor(target, options={}) {
+    /** The window to send and listen for messages with. */
+    this.target = target;
+
+    /**
+     * Set whether this frame acts as a server (fetches annotations from the
+     * API) or a client (contains annotatable content and displays highlights).
+     */
+    this.server = false;
+
+    /** Origins allowed to communicate with this frame. */
+    this.origin = '*';
+
+    /**
+     * Flag set in client frames to indicate when they are waiting for a
+     * confirmation from a server frame.
+     */
+    this.requestInProgress = false;
+
+    this.onDiscovery = null;
+
+    if (typeof options.server !== 'undefined') {
+      this.server = options.server;
+    }
+
+    if (typeof options.origin !== 'undefined') {
+      this.origin = options.origin;
+    }
+
+    this._onMessage = this._onMessage.bind(this);
+  }
+
+  /**
+   * Find other frames to communicate with.
+   *
+   * @param {DiscoveryCallback} onDiscovery - Callback to invoke with a token when
+   *   another frame is discovered.
+   */
+  startDiscovery(onDiscovery) {
+    if (this.onDiscovery) {
+      throw new Error('Discovery is already in progress. Call stopDiscovery() first');
+    }
+
+    this.onDiscovery = onDiscovery;
+
+    // Listen for messages from other frames.
+    this.target.addEventListener('message', this._onMessage, false);
+    this._beacon();
+  }
+
+  /**
+   * Stop listening for communication requests from other frames.
+   */
+  stopDiscovery() {
+    this.onDiscovery = null;
+    this.target.removeEventListener('message', this._onMessage);
+  }
+
+  /**
+   * Send a message to other frames in the current window to inform them about
+   * the existence of this frame and tell them whether this frame is a client
+   * or server.
+   */
+  _beacon() {
+    let beaconMessage;
+    if (this.server) {
+      beaconMessage = '__cross_frame_dhcp_offer';
+    } else {
+      beaconMessage = '__cross_frame_dhcp_discovery';
```

Comment:
These messages ought to be changed to something more obviously connected to Hypothesis. I left them as-is in this PR while I have a think about how the whole discovery process ought to work.